client/request.go

```diff
@@ -48,34 +49,44 @@ func makeRequestHandler(clientID uint32, seq sequenceGenerator, authen api.Authe
 // fetch corresponding Reply messages from.
 type requestSubmitter func(operation []byte) <-chan *messages.Reply
 
-// replyCollector collects f+1 matching Reply messages received from
-// the passed channel, finishes the request processing, and sends the
-// result of request execution to the passed channel.
-type replyCollector func(in <-chan *messages.Reply, out chan<- []byte)
+// requestRemover removes and stops further processing of the request
+// given its sequence number.
+type requestRemover func(seq uint64)
 
 // handleRequest initiates the specified operation to execute on the
 // replicated state machine using the passed request submitter and
 // returns a channel to receive the result of execution from.
-func handleRequest(operation []byte, submitter requestSubmitter, collector replyCollector) <-chan []byte {
+func handleRequest(operation []byte, submitter requestSubmitter, collector ReplyCollector, remover requestRemover) <-chan []byte {
 	resultChan := make(chan []byte, 1)
 	replyChan := submitter(operation)
-	go collector(replyChan, resultChan)
+	go collector.Collect(replyChan, resultChan, remover)
 	return resultChan
 }
 
 // makeReplyCollector constructs a replyCollector using the supplied
-// tolerance and request buffer to remove the request from when its
-// processing is finished.
-func makeReplyCollector(f uint32, buf *requestbuffer.T) replyCollector {
-	remover := makeRequestRemover(buf)
-	return func(in <-chan *messages.Reply, out chan<- []byte) {
-		collectReplies(f, in, remover, out)
+// tolerance to remove the request from when its processing is finished.
+func makeReplyCollector(f uint32) ReplyCollector {
+
+	return &replyCollector{
+
+		faults: f,
+		action: collectReplies,
```

Comment:
`collectReplies` could simply be a method attached to `replyCollector` type.

Reply:
Yeah, good point.